src/main/java/org/mockito/internal/invocation/MatcherApplicationStrategy.java

```diff
@@ -58,17 +58,24 @@ public static MatcherApplicationStrategy getMatcherApplicationStrategyFor(
      *         </ul>
      */
     public boolean forEachMatcherAndArgument(ArgumentMatcherAction action) {
-        if (invocation.getArguments().length == matchers.size()) {
-            return argsMatch(invocation.getArguments(), matchers, action);
-        }
-
         final boolean isVararg =
                 invocation.getMethod().isVarArgs()
-                        && invocation.getRawArguments().length == matchers.size()
-                        && isLastMatcherVarargMatcher(matchers);
+                        && invocation.getRawArguments().length == matchers.size();
 
         if (isVararg) {
-            int times = varargLength(invocation);
+            final Class<?> matcherType = lastMatcher().type();
+            final Class<?> paramType = lastParameterType();
+            if (paramType.isAssignableFrom(matcherType)) {
+                return argsMatch(invocation.getRawArguments(), matchers, action);
+            }
+        }
```

Comment:
Main matching logic change.

Similar to the other PR, but this time the code doesn't require the matcher to implement `VarargAware` matcher.